### PR TITLE
PoC: pagination support

### PR DESCRIPTION
**Pull request by pylipp**
This PR experiments with adding pagination support ~~on the data layer~~. I've worked with the `beneficiaries` resolvers, and added two optional arguments `cursor` (to indicate from where to start the query) and `limit` (indicate size of result).

- for `cursor` I've already updated the GraphQL schema
- for now, the cursor shall be of form `field:[start-value]` to indicate the field to paginate on, ~~and an optional start-value to start from. Later we can use encoded cursors, like e.g. the folks from [Slack](https://slack.engineering/evolving-api-pagination-at-slack/) (see section 'Encoded cursors')~~
- EDIT: use after/first instead of cursor/limit (see [Relay spec](https://relay.dev/graphql/connections.htm#sec-Forward-pagination-arguments))

### Next steps (done)

- compute pagination meta-info: at least `hasNextPage`, `nextCursor` for forward pagination
- return pagination meta-info in GraphQL response: requires type update like `Beneficiary` -> `BeneficiaryPage` which holds the result array, and a `PageInfo` struct

Let me know about any questions about the implementation, and whether you'd like other things to be experimented with :)

### CODE REVIEW

**Comment by pylipp** [comment]
[PoC: pagination support in data layer](https://trello.com/c/dTk5hA3X/781-poc-pagination-support-in-data-layer)

**Comment by pylipp** [comment]
Short sum-up from today's standup:
- paginate by id field only
- consider encoding/encrypting cursor (Slack uses base64 encoding, cf. link above; graphene by default in [node resolvers](https://docs.graphene-python.org/en/latest/relay/nodes/#quick-example)). @vahidbazzaz will share additional findings :)
- it would be 'nice-to-have' to not have to extend queries by the `cursor` and `limit` arguments, and instead pull those from the `info` argument of the resolver. However that'd result in a less clear implementation of the resolver

**Comment by vahidbazzaz** [comment]
@pylipp As far as I've found, there are common options for the cursor of choice: ID as the cursor, Timestamp and Opaque strings. But opaque strings appear to be popular, and base64 encode is frequently used.

> **ID as the cursor**
> API providers often choose a unique ID as the cursor value. For instance, the Twitter timeline APIs support tweet IDs as cursors. To fetch older tweets in a timeline, developers can pass the lowest ID received in the first set of results as the max_id parameter. The server then returns only tweets with IDs lower than or equal to the value of the max_id parameter.
> **Timestamp**
> Another common approach adopted by APIs returning time-based data, such as news feeds, is to use the timestamp as the cursor. Facebook APIs support until and since parameters, which accept Unix timestamps. When a timestamp is passed in the since request parameter, the Facebook API returns only items newer than the given timestamp.
> **Opaque strings**
> Using opaque strings as the cursor is increasingly becoming the preferred choice for API providers. Although they appear as random sets of characters, they are generally encoded values. A key advantage of using opaque strings is the ability to encode additional information within a single cursor. Large-scale applications can encode multiple IDs or an ID-plus-pointer combination to a database shard in these cursor values. Modern versions of various APIs, including those of Slack, Facebook, GitHub, and Twitter, use opaque strings as cursors.

**Comment by pylipp** [comment]
Thanks @vahidbazzaz , that appears to be quite a comprehensive resource you found! I'll add the base64 encoding then.

About 'abstracting away' the cursor and limit arguments I would not go for using the `info` object. From ariadne discussions I read that parsing the underlying GraphQLResolveInfo object is expensive, and we want the resolver functions to be fast and direct w.r.t. the schema. Though I think it makes sense to have something like a `PaginationInput` type to group the pagination arguments. Especially once we also support client-side filtering in queries a form like `beneficiaries(paginationInput, filterInput)` looks more structured IMHO than having all arguments written out flat. What do you think about `PaginationInput`?

**Comment by vahidbazzaz** [comment]
> Thanks @vahidbazzaz , that appears to be quite a comprehensive resource you found! I'll add the base64 encoding then.
> 
> About 'abstracting away' the cursor and limit arguments I would not go for using the `info` object. From ariadne discussions I read that parsing the underlying GraphQLResolveInfo object is expensive, and we want the resolver functions to be fast and direct w.r.t. the schema. Though I think it makes sense to have something like a `PaginationInput` type to group the pagination arguments. Especially once we also support client-side filtering in queries a form like `beneficiaries(paginationInput, filterInput)` looks more structured IMHO than having all arguments written out flat. What do you think about `PaginationInput`?

Thanks @pylipp I think the `PaginationInput` makes perfect sense. :)

**Review by vahidbazzaz — APPROVED** [review]

File: flask/boxwise_flask/graph_ql/resolvers.py [hunk 1]
```diff
@@ -159,9 +176,16 @@ def resolve_products(_, info):
 
 
 @query.field("beneficiaries")
-def resolve_beneficiaries(_, info):
+def resolve_beneficiaries(_, info, cursor=None, limit=50):
```
[the remark this answers is not in the document]
Comment (pylipp) [reply flask/boxwise_flask/graph_ql/resolvers.py]: Sorry, it seems like I didn't push the latest changes in time. The beneficiaries resolver now takes a `PaginationInput` argument. I would still vote for explicitly extending the definitions of those queries that we want to support pagination with.

File: flask/boxwise_flask/graph_ql/resolvers.py [hunk 2]
```diff
@@ -38,6 +38,23 @@
...
+    ModelSelect.where() clause for the given model.
+    For now, this expects the name of an existing model field (default: 'id') as cursor.
+    Optionally, an integer offset can be specified after a colon.
+    """
+    offset = 0
+    if cursor is None:
+        field = "id"
+    else:
+        try:
+            field, offset = cursor.split(":")
```
[the remark this answers is not in the document]
Comment (pylipp) [reply flask/boxwise_flask/graph_ql/resolvers.py]: This is true, has been updated in the meantime!

**Comment by pylipp** [comment]
@vahidbazzaz the `beneficiaries` queries now returns a `BeneficiaryPage` result. Under the hood, I've sorted the pagination-specific functions into a separate module.

I feel like this is now very close to being completed :sweat_smile: I could include a `totalCount` field for the `BeneficiaryPage`, what do you think? Let me know if you have any other points to address.

**Comment by pylipp** [comment]
[Research DB Design and Flask/Python/Peewee modules that would enable pagination for long lists of records #GraphQL](https://trello.com/c/3meTtCGt/728-research-db-design-and-flask-python-peewee-modules-that-would-enable-pagination-for-long-lists-of-records-graphql)

**Comment by vahidbazzaz** [comment]
> @vahidbazzaz the `beneficiaries` queries now returns a `BeneficiaryPage` result. Under the hood, I've sorted the pagination-specific functions into a separate module.
> 
> I feel like this is now very close to being completed 😅 I could include a `totalCount` field for the `BeneficiaryPage`, what do you think? Let me know if you have any other points to address.

@pylipp Having read the Relay spec, I wonder if we should also implement [backward pagination arguments](https://relay.dev/graphql/connections.htm#sec-Backward-pagination-arguments) e.g.` last-before`?

**Comment by pylipp** [comment]
@vahidbazzaz yes, I think so :) I created https://trello.com/c/XwpRArzY/785-graphql-pagination-follow-up (also FYI @aerinsol ) to not blow up the present PR too much :sweat_smile:

**Pull request merged**